User Preference to Hide beat grid on waveform

williamlemus opened the pull request:
These changes are to enable the functionality requested in bug 1659888 : https://bugs.launchpad.net/mixxx/+bug/1659888 . I have done some testing on it and it is working. Please let me know if anything needs to be changed.

	modified:   src/preferences/dialog/dlgprefwaveform.h
	modified:   src/preferences/dialog/dlgprefwaveformdlg.ui
	modified:   src/waveform/renderers/waveformrenderbeat.cpp
	modified:   src/waveform/renderers/waveformwidgetrenderer.cpp
	modified:   src/waveform/renderers/waveformwidgetrenderer.h
	modified:   src/waveform/waveformwidgetfactory.cpp
	modified:   src/waveform/waveformwidgetfactory.h

Added functionality to hide beat grid lines on waveform.

williamlemus committed:
modified:   src/preferences/dialog/dlgprefwaveform.cpp
	modified:   src/preferences/dialog/dlgprefwaveform.h
	modified:   src/preferences/dialog/dlgprefwaveformdlg.ui
	modified:   src/waveform/renderers/waveformrenderbeat.cpp
	modified:   src/waveform/renderers/waveformwidgetrenderer.cpp
	modified:   src/waveform/renderers/waveformwidgetrenderer.h
	modified:   src/waveform/waveformwidgetfactory.cpp
	modified:   src/waveform/waveformwidgetfactory.h
Added functionality to hide beat grid lines on waveform.

daschuer commented on src/waveform/renderers/waveformwidgetrenderer.cpp:
```diff
@@ -274,6 +275,10 @@ void WaveformWidgetRenderer::setZoom(int zoom) {
 }
 
+void WaveformWidgetRenderer::setDisplayGrid(bool set){
+    m_renderGrid = set;
```
Render or display?

daschuer commented on src/waveform/waveformwidgetfactory.cpp:
```diff
@@ -223,6 +224,13 @@ bool WaveformWidgetFactory::setConfig(UserSettingsPointer config) {
         m_config->set(ConfigKey("[Waveform]","ZoomSynchronization"), ConfigValue(m_zoomSync));
     }
 
+    int displayGrid = m_config->getValueString(ConfigKey("[Waveform]", "beatGridLinesCheckBox")).toInt(&ok);
```
Later we may use other controls, but the config key remains. 
how about just "showBeatgrid"? 

By the way, you can use here the m_config->getValue() template function.

daschuer commented on src/waveform/waveformwidgetfactory.cpp:
```diff
@@ -432,6 +440,22 @@ void WaveformWidgetFactory::setZoomSync(bool sync) {
     }
 }
 
+void WaveformWidgetFactory::setDisplayGrid(bool sync){
```
Use BeatGrid, to be free to introduce other type of grids.

williamlemus replied:
Ok. I will update all the function to better reflect that we are referring to the Beat Grid.

daschuer commented:
Thank you for adopting this. I have just left some renaming suggestions. 
I hope I find time to test this soon.

williamlemus committed:
Modified naming to better reflect functionality.

Be-ing commented on src/waveform/waveformwidgetfactory.cpp:
```diff
@@ -432,6 +440,22 @@ void WaveformWidgetFactory::setZoomSync(bool sync) {
+        m_config->set(ConfigKey("[Waveform]", "beatGridLinesCheckBox"), ConfigValue(m_beatGridEnabled));
+    }
+
+    if(m_waveformWidgetHolders.size() == 0){
```
nitpick: please use spaces around conditional expressions:
`if (m_waveformWidgetHolders.size() == 0) {`

williamlemus committed:
Adding correct spacing around curly braces

Be-ing commented:
Thanks for working on this. I tested it and it works until Mixxx is restarted. If the option is unchecked (beatgrid is hidden) on shutdown, when Mixxx starts again, the beatgrid will be back even the the option in the preferences is unchecked.

Be-ing commented on src/preferences/dialog/dlgprefwaveformdlg.ui:
```diff
@@ -330,6 +343,16 @@
+     <item row="12" column="1">
+      <widget class="QCheckBox" name="beatGridLinesCheckBox">
+       <property name="toolTip">
+        <string>Display beat grid lines on waveform</string>
```
Change this to just "Beat grid"

williamlemus replied:
I will make these changes, and look into why the preference is being ignored when mixxx is restarted.

williamlemus replied:
@Be-ing  I've resolved the issue with the preference being ignored when starting Mixxx. Which lines am I changing to "Beat Grid" and "Show Beat Markers"? I'll push all the changes together.

Be-ing commented on src/preferences/dialog/dlgprefwaveformdlg.ui:
```diff
@@ -330,6 +343,16 @@
+        <string>Display beat grid lines on waveform</string>
+       </property>
+       <property name="text">
+        <string>Enable Beat Grid Lines</string>
```
Change this to "Show beat markers"

Be-ing commented:
Ping @williamlemus, have you continued work on this? I'd like to increase the maximum waveform zoom out level further after merging this. The reason I chose the limit I did in #1165 is because it started to look weird when zoomed out far and showing the beat markers.

williamlemus commented:
@Be-ing Hey I haven't had a chance to figure out how to get around the issue where it's defaulting. I'll try to take a look by mid week and reach out via IRC if I'm still stuck.

williamlemus committed:
Resolved issue with beat grid preference being ignored upon starting up mixxx

williamlemus commented:
@Be-ing I've made the updates requested. Please let me know if anything else needs to be corrected.

daschuer commented on src/waveform/waveformwidgetfactory.cpp:
```diff
@@ -223,6 +224,13 @@ bool WaveformWidgetFactory::setConfig(UserSettingsPointer config) {
         m_config->set(ConfigKey("[Waveform]","ZoomSynchronization"), ConfigValue(m_zoomSync));
     }
 
```
you can simplify the code passing the default value

int showBeatGrid = m_config->getValue("[Waveform]", "beatGridLinesCheckBox", m_showBeatGrid)

(not sure about the exact syntax though)

daschuer commented on src/waveform/waveformwidgetfactory.cpp:
```diff
@@ -432,6 +441,22 @@ void WaveformWidgetFactory::setZoomSync(bool sync) {
+void WaveformWidgetFactory::setDisplayBeatGrid(bool sync) {
+    m_beatGridEnabled = sync;
+    if (m_config) {
+        m_config->set(ConfigKey("[Waveform]", "beatGridLinesCheckBox"), ConfigValue(m_beatGridEnabled));
```
we normally do not set the value, if not required. We try to load the value and if it fails pick the default. Is it required here?

williamlemus replied:
The dlgprefwaveform calls this in order to update the Config
`void DlgPrefWaveform::slotSetGridLines(bool displayGrid) {
    WaveformWidgetFactory::instance()->setDisplayBeatGrid(displayGrid);
}`

Would there be a better way to do this?

Be-ing replied:
I think it would make more sense to set the ConfigKey in DlgPrefWaveform than here.

Be-ing replied:
Oh, I see the way you did it is consistent with the surrounding code. You don't have to clean this up if you don't have time, but it would be nice to move the setting of the ConfigKeys into DlgPrefWaveform.

williamlemus replied:
Should I do this just for the display grid preference?

Be-ing replied:
No, keep it consistent. Either leave it how it is or refactor how all the settings in this function are set.

Be-ing commented:
I tested again and it works as expected. I think the UI in the Preferences could be improved. I think it would be difficult for users to discover on their own all the way at the bottom of the preference page. How about putting it between "End of track warning" and "Default zoom level"? I think that will make it more obvious to users that the option exists if they choose to zoom the waveform out far and see the grid lines close together and cluttering the view. For the wording, I suggest changing it to "Beat grid" for the label in the left column and "Show beat markers" for the checkbox label in the right column.

esbrandt commented:
Thanks for adding the feature.

If you plan to extend this any further, adding a ``Dim`` mode can be an option.
The dimmed mode is basically the normal mode, just with lower alpha ``0.3`` instead of ``0.9`` for the [beat grid lines](https://github.com/mixxxdj/mixxx/blob/master/src/waveform/renderers/waveformrenderbeat.cpp#L27).
Traktor has this option and it is really neat. Gives access to the bead grid without obstructing the view on the waveform/markers as much as with the default beatgrid.

williamlemus commented:
@esbrandt That sounds like a good idea. Can I do the dimming functionality as a separate PR? I would like to get this PR as I'm not sure if I'll have time the next couple of weeks to extend it.

esbrandt commented:
> @esbrandt That sounds like a good idea. Can I do the dimming functionality as a separate PR? I would like to get this PR as I'm not sure if I'll have time the next couple of weeks to extend it.

Totally up to you.
Just submit a new PR when you are ready.

williamlemus committed:
Simplify code to get default value in waveform widget. Move Beat Grid check box for easier discover

williamlemus commented:
I've simplified the code as requested. @Be-ing  I also moved up the beat grid option and changed the text.

Please let me know if I missed anything.

Be-ing commented on src/waveform/waveformwidgetfactory.cpp:
```diff
@@ -224,12 +224,9 @@ bool WaveformWidgetFactory::setConfig(UserSettingsPointer config) {
-    } else {
-        m_config->set(ConfigKey("[Waveform]", "beatGridLinesCheckBox"), ConfigValue(m_beatGridEnabled));
-    }
+    int showBeatGrid = m_config->getValue(ConfigKey("[Waveform]", "beatGridLinesCheckBox"), m_beatGridEnabled);
```
If you change this to a bool you don't need to do the static_cast on the next line.

Be-ing commented:
The order of the items in the preferences looks odd because there are 3 checkboxes in the right column but only a label in the left column for the middle one. I think moving the new beatgrid line preference between "End of track warning" and "Default zoom level" would work better.
![waveform-preferences](https://user-images.githubusercontent.com/9455094/28252758-28be6858-6a5f-11e7-9611-1606726b75db.png)

williamlemus committed:
Update position of Beat grid checkbox. Cleared up code set beat grid.

Be-ing commented:
Looks good to me!

williamlemus commented:
Should I worry about the unsuccessful check? Also, I'll start working on the dimming, and will make a PR to adding dimming and move the set config into waveform preferences.

Be-ing commented:
No, we use the free plans for Travis and AppVeyor that have short time limits. Randomly builds will run out of time. AppVeyor passed, so it's okay.|

> Also, I'll start working on the dimming, and will make a PR to adding dimming and move the set config into waveform preferences.

Cool, thanks.

daschuer commented:
Thank you, @williamlemus and all involved. :-)
